src/renderers/dom/client/ReactMount.js

```diff
@@ -468,6 +469,14 @@ var ReactMount = {
     var containerHasNonRootReactChild = hasNonRootReactChild(container);
 
     if (__DEV__) {
+      var validElement = typeof nextElement.type === 'string' ?
+      nextElement.type.match(/\-|\s+/) === null &&
+      document.createElement(nextElement.type) instanceof HTMLUnknownElement &&
+      !ReactDOMFactories.hasOwnProperty(nextElement.type) : false;
+
+      warning(!validElement, 'ReactDOM.render(): Invalid component element. ' +
+        'Make sure the component starts with an upper-case letter.');
```

Comment:
Agree! I was not too happy with the message itself! Thanks for the suggestion, I'll change it to that.